core/arch/arm/kernel/secure_partition.c

```diff
@@ -299,7 +301,28 @@ TEE_Result sp_map_shared(struct sp_session *s,
 	if (*va)
 		return TEE_ERROR_NOT_SUPPORTED;
 
+	if (is_device) {
+		uint8_t dev_type = (smem->mem_reg_attr >> FFA_DEV_MEM_SHIFT) &
+				    FFA_DEV_MEM_MASK;
+
+		if (sp_mem_ffa_dev_to_cache_type(dev_type, &cache_type))
+			return FFA_INVALID_PARAMETERS;
+	}
+
 	SLIST_FOREACH(reg, &smem->regions, link) {
+		/* Check device region attributes */
+		if (is_device) {
+			uint32_t mcache_type = 0;
+
+			if (mobj_get_cattr(reg->mobj, &mcache_type))
+				return FFA_INVALID_PARAMETERS;
+
+			if (cache_type != mcache_type) {
```

Comment:
Couldn't this check have been performed earlier when the smem was created? It seems that the smem is in an inconsistent state that never should have been allowed if this test fails.

Reply:
This check is not needed like you are saying, smem is the one which is setting the cache_type before.